aspnetcore/tutorials/ai-powered-group-chat/ai-powered-group-chat.md

```diff
@@ -0,0 +1,110 @@
+---
+
+## Implementation
+
+In this section, we'll walk through the key parts of the code that integrate SignalR with OpenAI to create an AI-enhanced group chat experience.
+
+### Data flow 
+
+:::image type="content" source="./sequence-diagram-ai-powered-group-chat.png" alt-text="sequence diagram for the AI-powered group chat":::
+
+### SignalR Hub integration
+
+The `GroupChatHub` class manages user connections, message broadcasting, and AI interactions. When a user sends a message starting with `@gpt`, the hub forwards it to OpenAI, which generates a response. The AI's response is streamed back to the group in real-time.
+```csharp
+var chatClient = _openAI.GetChatClient(_options.Model);
+
+await foreach (var completion in 
+    chatClient.CompleteChatStreamingAsync(messagesInludeHistory))
+{   
+    // ...
+    // Buffering and sending the AI's response in chunks
+    await Clients.Group(groupName).SendAsync(
+        "newMessageWithId",
+        "ChatGPT",
+        id,
+        totalCompletion.ToString());
+    // ...
+}
+```
+
+### Maintain context with history
+
+Every request to [Open AI's Chat Completions API](https://platform.openai.com/docs/guides/chat-completions) is stateless—Open AI doesn't store past interactions. In a chat app, what a user or an assistant has said is important for generating a response that's contextually relevant. To achieve this, include chat history in every request to the Completions API. 
+
+The `GroupHistoryStore` class manages chat history for each group. It stores messages posted by both the users and AI assistants, ensuring that the conversation context is preserved across interactions. This context is crucial for generating coherent AI responses.
+
+```csharp
+// Store message generated by AI-assistant in memory
+public void UpdateGroupHistoryForAssistant(string groupName, string message)
+{
+    var chatMessages = _store.GetOrAdd(groupName, _ => InitiateChatMessages());
+    chatMessages.Add(new AssistantChatMessage(message));
+}
+```
+
+```csharp
+// Store message generated by users in memory
+_history.GetOrAddGroupHistory(groupName, userName, message);
+```
+
+### Stream AI responses
+
+The `CompleteChatStreamingAsync()` method streams responses from OpenAI incrementally, which allows the app to send partial responses to the client as they are generated. 
+
+The code uses a `StringBuilder` to accumulate the AI's response. It checks the length of the buffered content and sends it to the clients when it exceeds a certain threshold (e.g., 20 characters). This approach ensures that users see the AI’s response as it forms, mimicking a human-like typing effect. 
+```csharp
+totalCompletion.Append(content);
+
+if (totalCompletion.Length - lastSentTokenLength > 20)
+{
+    await Clients.Group(groupName).SendAsync(
+        "newMessageWithId",
+        "ChatGPT",
+        id,
+        totalCompletion.ToString());
+
+    lastSentTokenLength = totalCompletion.Length;
```

Comment:
nit: This is something I don't like about bing chat, it sends the entire message every update. Really it should just append the new content, saves bandwidth and memory usage on the server.